### PR TITLE
Adjust user input date to UTC based on project timezone

### PR DESCRIPTION
**Pull request by minhaminha**
## Product Description

This PR fixes a longstanding issue wherein dates that users input into the query search box aren't translated back to UTC from the domain's timezones, which can cause cases outside the specified date range to be returned or cases inside the range be hidden. This fix ONLY applies to the Case List Explorer page. 

## Technical Summary

[Jira Ticket](https://dimagi-dev.atlassian.net/browse/SAAS-14443)

All this does it use `adjust_input_date_by_timezone` to convert the raw date value to UTC based on the domain's timezone while also adjusting for certain comparisons where the date needs to be moved up a date to be inclusive of the original date.  

This is one of the several areas identified wherein user input dates or datespans are not being translated back to UTC based on the user's or their domain's timezones. 

## Feature Flag

Case List Explorer

### Safety story

Tested locally and on staging. QA ongoing. 

### Automated test coverage

Added tests related to user input being correctly translated to UTC. 

### QA Plan

[QA ticket](https://dimagi-dev.atlassian.net/browse/QA-5159)

### Rollback instructions

- [x] This PR can be reverted after deploy with no further considerations

### Labels & Review
- [x] Risk label is set correctly
- [x] The set of people pinged as reviewers is appropriate for the level of risk of the change

### CODE REVIEW

File: corehq/apps/case_search/xpath_functions/comparison.py [hunk 1]
```diff
@@ -1,12 +1,17 @@
+from datetime import datetime, timedelta
+
 from django.utils.translation import gettext
 from eulxml.xpath import serialize
 from eulxml.xpath.ast import Step
 
 from corehq.apps.case_search.dsl_utils import unwrap_value
-from corehq.apps.case_search.exceptions import CaseFilterError
+from corehq.apps.case_search.exceptions import CaseFilterError, XPathFunctionException
+from corehq.apps.case_search.xpath_functions.value_functions import _value_to_date
```
Comment (biyeun) [inline corehq/apps/case_search/xpath_functions/comparison.py:9]: nit: it's not the best practice to import private methods (starting with `_`). I think an argument can be made to refactor the name of this method to a public one `value_to_date`.

File: corehq/apps/es/tests/test_case_search_es.py [hunk 2]
```diff
@@ -543,3 +545,80 @@ def test_starts_with_query(self):
             "starts-with(ssn, '100')",
             ['c5', 'c6', 'c2']
         )
+
+    def test_user_input_forward_timezone_adjustment(self):
+        """
```
Comment (biyeun) [inline corehq/apps/es/tests/test_case_search_es.py:550]: love these tests and descriptions!

Comment (biyeun) [reply corehq/apps/es/tests/test_case_search_es.py:550]: the only nit I have is that each assertion / scenario should be its own test. that way if one of the scenarios fail at the beginning, it doesn't cause the remaining scenarios to fail by default

**Review by biyeun — APPROVED** [review]
looks good! just some nits that are not blocking on my end.

**Comment by millerdev** [comment]
Is this related to https://github.com/dimagi/commcare-hq/pull/33056? cc @mjriley

**Comment by minhaminha** [comment]
> Is this related to #33056? cc @mjriley

These changes are pretty isolated from what Matt's doing, which I think is limited to datespans received by the datespan widget thing, whereas my changes are specific to case search / CLE and adjusting individual dates the user inputs. Though this does bring up the point again that timezone handling in situations like these could definitely be standardized across hq. @millerdev

File: corehq/apps/case_search/xpath_functions/comparison.py [hunk 3]
```diff
@@ -18,18 +23,35 @@ def property_comparison_query(context, case_property_name_raw, op, value_raw, no
...
+                                             **{RANGE_OP_MAPPING[op]: value})
         except (TypeError, ValueError):
             raise CaseFilterError(
                 gettext("The right hand side of a comparison must be a number or date. "
                   "Dates must be surrounded in quotation marks"),
                 serialize(node),
             )
+
+
+def adjust_input_date_by_timezone(date, timezone, op):
```
Comment (Jtang-1) [inline corehq/apps/case_search/xpath_functions/comparison.py:53]: This change seems to have caused this [bug](https://dimagi-dev.atlassian.net/jira/servicedesk/projects/SUPPORT/queues/custom/81/SUPPORT-17192) for date case properties. I.E if the xpath query is`dob="1984-08-15"`, "1984-08-15" is being converted to some other value even though dob is stored in ES as "1984-08-15". The filter is built with the converted value, leading to no matches.

I've been thinking about what the fix would be and i'm a bit stumped. The only thing i've thought of: to know whether to apply to conversion, we need to know if the property being filtered is stored as UTC. A way to do that would be to only convert `CommCareCase` fields of type `DateTimeField`. But that doesn't sounds like a great strategy. 
Commenting to see if anything else might come to mind that I'm not thinking of.

Comment (minhaminha) [reply corehq/apps/case_search/xpath_functions/comparison.py:53]: Sorry I didn't get a chance to look at this properly on Friday but yea this is absolutely my bad. The error I was tackling described date ranges so testing the equal operator somehow never crossed my mind. That being said, I think a simple fix off the top of my head might be just to put a conditional that says ignore the timezone conversion logic if '=' is present. Not the most elegant thing but I'd be happy to work with you more on this.

Comment (minhaminha) [reply corehq/apps/case_search/xpath_functions/comparison.py:53]: Oh nevermind I just saw the other PR

Comment (Jtang-1) [reply corehq/apps/case_search/xpath_functions/comparison.py:53]: Thanks Minha! I should have updated this comment earlier - https://github.com/dimagi/commcare-hq/pull/33272

**Pull request merged**